### PR TITLE
VIR-36: Vacancy number

### PR DESCRIPTION
**Pull request by Peskarhu**
* Add utility for generating vacancy numbers using cost center (kustannuspaikka) and sequential number (i.e. 4120 0001).
* Allow vacancy number to be null due to current database issues...might be my skill issue :(
* Generate and save vacancy number if null when fetching all positions or creating new one.
* Add vacancy number check when creating and fetching positions (may need refactoring).
* Update UI to display vacancy number in table, including translations.

### CODE REVIEW

**Comment by Peskarhu** [comment]
VIR-36_test

File: backend/Virkarekisteri/src/Utils/VacancyNumberGenerator.cs
Comment (0x464e) [inline backend/Virkarekisteri/src/Utils/VacancyNumberGenerator.cs:1]: We can implement the logic for getting the org number and latest vacancy number in the position repository.

In this case we don't need to inject the database into this file, instead you'd inject the position repository and just use two simple function calls. (Bonus: when injecting the position repository, you can use a primary constructor. See e.g. how it's done in GetPosition function)

(So in the PositionRepository you'd create two new functions, called something like GetOrgNumberById and GetLatestVacancyNumberByPrefix, and then use the two functions here)

File: backend/Virkarekisteri/src/Utils/VacancyNumberGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+            .OrganizationTrees.Where(o => o.Id == orgTreeId)
+            .Select(o => o.Number)
+            .FirstOrDefaultAsync();
+
+        string vacancyPrefix = orgNumber ?? "";
+
+        // Get the most recent vacancy number with the same prefix if there is
+        var latestVacancyNumber = await db
+            .Positions.Where(p => p.VacancyNumber.StartsWith(vacancyPrefix))
+            .OrderByDescending(p => p.VacancyNumber)
```
Comment (0x464e) [inline backend/Virkarekisteri/src/Utils/VacancyNumberGenerator.cs]: VacancyNumber is a string, is this ordering going to work as expected in all cases?

File: backend/Virkarekisteri/src/Utils/VacancyNumberGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        var orgNumber = await db
+            .OrganizationTrees.Where(o => o.Id == orgTreeId)
+            .Select(o => o.Number)
+            .FirstOrDefaultAsync();
+
+        string vacancyPrefix = orgNumber ?? "";
+
+        // Get the most recent vacancy number with the same prefix if there is
+        var latestVacancyNumber = await db
+            .Positions.Where(p => p.VacancyNumber.StartsWith(vacancyPrefix))
```
Comment (0x464e) [inline backend/Virkarekisteri/src/Utils/VacancyNumberGenerator.cs]: Compiler warns that we're referencing a possible null value here (p.VacancyNumber can be null)

You can check for null before accessing the value:  
`.Positions.Where(p => p.VacancyNumber != null && p.VacancyNumber.StartsWith(vacancyPrefix))`

**Review by 0x464e — CHANGES_REQUESTED** [review]

File: backend/Virkarekisteri/src/Utils/VacancyNumberGenerator.cs (hunk 1 above)
Comment (Peskarhu) [reply backend/Virkarekisteri/src/Utils/VacancyNumberGenerator.cs]: Should work as vacancyNumber will be 8 char long for every case (4 prefix + 4 sequence) and the sequence is zero padded. This was discussed that we will now only have kustannuspaikka which is 4 long

File: backend/Virkarekisteri/src/Program.cs [hunk 3]
```diff
@@ -4,6 +4,7 @@
 using Microsoft.Extensions.Hosting;
 using Virkarekisteri.Models;
 using Virkarekisteri.Repositories;
+using Virkarekisteri.Utils;
```
Comment (0x464e) [inline backend/Virkarekisteri/src/Program.cs:7]: Forgotten import here, but can be fixed in your next PR, whatever it may be.

**Review by 0x464e — APPROVED** [review]

File: backend/Virkarekisteri/src/Program.cs (hunk 3 above)
Comment (Peskarhu) [reply backend/Virkarekisteri/src/Program.cs:7]: True

**Pull request merged**